alf/algorithms/prior_actor.py

```diff
@@ -0,0 +1,282 @@
+from alf.data_structures import AlgStep, Experience, TimeStep, StepType
+from alf.networks import Network
+from alf.tensor_specs import TensorSpec, BoundedTensorSpec
+
+
+def normcdf(a, b):
+    # 0.7071067811865476 = math.sqrt(0.5)
+    return 0.5 * (
+        torch.erf(0.7071067811865476 * b) - torch.erf(0.7071067811865476 * a))
+
+
+class TruncatedNormal(td.Distribution):
+    def __init__(self, loc, scale, low, high, validate_args=None):
+        """Normal distribution truncated to the range between ``low`` and ``high``.
+
+        Currently, only ``log_prob()`` is implemented.
+
+        Args:
+            loc (Tensor): mean of the untruncated Normal
+            scale (Tensor): standard deviation of the untruncated Normal
+            low (Tensor): lower range of the truncation range
+            high (Tensor): upper range of the truncation range
+        """
+        self._loc = loc
+        self._scale = scale
+        self._low = low
+        self._high = high
+        super().__init__(batch_shape=loc.shape, validate_args=validate_args)
+
+    def log_prob(self, value):
+        """Log-probability of ``value``.
+
+        Args:
+            value (Tensor): the samples whose log_prob is to calculated
+        Returns:
+            log probability of ``value``
+        """
+        scale = self._scale
+        loc = self._loc
+        var = scale**2
+        log_scale = scale.log()
+        low = (self._low - loc) / scale
+        high = (self._high - loc) / scale
+        # 0.9189385332046727 = math.log(math.sqrt(2 * math.pi))
+        log_prob = -(
+            (value - loc)**2) / (2 * var) - log_scale - 0.9189385332046727
+        return log_prob - normcdf(low, high).log()
+
+    def sample(self):
+        raise NotImplementedError()
+
+    def rsample(self):
+        raise NotImplementedError()
+
+
+class MixtureSameFamily(td.Distribution):
+    """
+    ``MixtureSameFamily`` was introduced in pytorch 1.5. Here we only implement
+    a subset of its functions.
+    """
+
+    def __init__(self,
+                 mixture_distribution,
+                 component_distribution,
+                 validate_args=None):
+        self._mixture_distribution = mixture_distribution
+        self._component_distribution = component_distribution
+        event_shape = component_distribution.event_shape
+        self._event_ndims = len(event_shape)
+        super().__init__(
+            batch_shape=component_distribution.batch_shape[:-1],
+            event_shape=event_shape,
+            validate_args=validate_args)
+
+    def log_prob(self, x):
+        """Log-probability of sample ``x``."""
+        x = x.unsqueeze(-1 - self._event_ndims)
+        log_prob_x = self._component_distribution.log_prob(x)  # [S, B, k]
+        log_mix_prob = self._mixture_distribution.logits  # [B, k]
+        return torch.logsumexp(log_prob_x + log_mix_prob, dim=-1)  # [S, B]
+
+
+@gin.configurable
+class PriorActor(Algorithm):
```

Comment:
The name of this class could be better. Right now it sounds like a general class of "uniformprioractor", but it's actually quite different.